File: Object-C/BlackFireIM/Example/Lib/MSIMSDK-UIKit/ChatDetail/View/Message/MSBubbleMessageCell.m
//
//  MSBubbleMessageCell.m
//  BlackFireIM
//
//  Created by benny wang on 2021/3/5.
//

#import "MSBubbleMessageCell.h"
#import "UIView+Frame.h"

@implementation MSBubbleMessageCell

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        _bubbleView = [[UIImageView alloc] initWithFrame:CGRectZero];
        [self.container addSubview:_bubbleView];
        _bubbleView.bounds = self.container.bounds;
        _bubbleView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    }
    return self;
}

- (MSBubbleMessageCellData *)bubbleData
{
    return (MSBubbleMessageCellData *)self.messageData;
}

- (void)fillWithData:(MSBubbleMessageCellData *)data
{
    [super fillWithData:data];
    self.bubbleView.image = data.bubble;
    self.bubbleView.highlightedImage = data.highlightedBubble;
}

- (void)layoutSubviews
{
    [super layoutSubviews];
}

@end

File: Swift/MSIM-Swift/Lib/MSIMSDK-UIKit/ChatDetail/View/Message/MSBubbleMessageCell.swift
//
//  MSBubbleMessageCell.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/8/6.
//

import UIKit

open class MSBubbleMessageCell: MSMessageCell {

    public var bubbleView: UIImageView!
    
    public var bubbleData: MSBubbleMessageCellData?
    
    public override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        
        bubbleView = UIImageView(frame: .zero)
        container.addSubview(bubbleView)
        bubbleView.bounds = container.bounds
        bubbleView.autoresizingMask = [.flexibleWidth,.flexibleHeight]
        
    }
    
    required public init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    public override func fillWithData(data: MSMessageCellData) {
        super.fillWithData(data: data)
        if let bubbleData = data as? MSBubbleMessageCellData {
            self.bubbleData = bubbleData
            bubbleView.image = bubbleData.bubble
            bubbleView.highlightedImage = bubbleData.highlightedBubble
        }
    }
}
